File: src/model_metric.cc
#include <iostream>

#include "counter/model_metric.h"
using namespace std;

ModelMetric* ModelMetric::get_instance() {
  if (!modelmetric) {
    modelmetric = new ModelMetric();
    if (modelmetric) {
      modelmetric->thrd = new std::thread(&ModelMetric::process_queue, modelmetric);
    }
  }
  return modelmetric;
};

ModelMetric::~ModelMetric() {
  delete thrd;
  thrd = nullptr;
};

void ModelMetric::process_queue() {
  while (!stop_process) {
    std::this_thread::sleep_for(std::chrono::seconds(CALL_HOME_MODEL_RUN_COUNT_TIME_SECS));
    std::map<std::string, int> dict = ModelExecCounter::get_dict();
    for(auto pair_dict :  dict) {
      char buff[128];
      snprintf(buff, sizeof(buff), "{ \"record_type\": %s, \"model\": \"%s\", \"uuid\": \"%s\", \"run_count\": \"%s\" }", std::to_string(MODEL_RUN).c_str(), pair_dict.first.c_str(), device_id.c_str(), std::to_string(pair_dict.second).c_str());
      std::string pub_data = buff;
      publisher->send(pub_data);
    }
  }
}

ModelMetric* ModelMetric::modelmetric = nullptr;

File: python/dlr/counter/model_metric.py
"""Model Metric Publisher"""
import concurrent.futures
import logging
import json
import threading


from . import config
from .model_exec_counter import ModelExecCounter
from .publisher import MsgPublisher


class ModelMetric(object):
    """Model metric publisher class"""
    _pub_model_metric = True
    _instance = None
    MODEL_RUN = 3

    @staticmethod
    def get_instance(uuid):
        """return single instance of class"""
        if ModelMetric._instance is None:
            ModelMetric._instance = ModelMetric(uuid)
        return ModelMetric._instance

    def __init__(self, uuid=0):
        try:
            self.uuid = uuid
            self.publisher = MsgPublisher.get_instance()
            # start loop
            self.executor = concurrent.futures.ThreadPoolExecutor(max_workers=1)
            self.condition = threading.Condition()
            self.executor.submit(self.push_model_metric, self.condition)
            logging.info("model metric thread pool execution started")
        except Exception:
            logging.exception("model metric thread pool not started", exc_info=False)

    def push_model_metric(self, condv):
        """publishing model run metric"""
        while ModelMetric._pub_model_metric:
            with condv:
                condv.wait(config.CALL_HOME_MODEL_RUN_COUNT_TIME_SECS)
                mod_dict = ModelExecCounter.get_dict()
                if mod_dict:
                    for key, val in mod_dict.items():
                        self.model_run_info_publish(ModelMetric.MODEL_RUN, key, val)

    def model_run_info_publish(self, model_event_type, model, count=0):
        """push model load information at time model load time"""
        try:
            pub_data = {'record_type': model_event_type, 'model': model, 'uuid': self.uuid, 'run_count': count}
            self.push(pub_data)
        except Exception:
            logging.exception("unable to complete model count", exc_info=False)

    def push(self, data):
        """publish information to Server"""
        self.publisher.send(json.dumps(data))

    def stop(self):
        """stop  thread execution"""
        with self.condition:
            self.condition.notify_all()
            ModelMetric._pub_model_metric = False
            self.executor.shutdown(wait=False)
